File: starter/java/src/main/java/kata/Message.java
package kata;

import java.time.Duration;
import java.time.Instant;
import java.util.Comparator;

public class Message {

  private final User userWhoPostedMessage;
  private final String details;
  private final Instant timePosted;

  public Message(User userWhoPostedMessage, String details) {
    this.userWhoPostedMessage = userWhoPostedMessage;
    this.details = details;
    this.timePosted = Instant.now();
  }

  public User getUserWhoPostedMessage() {
    return userWhoPostedMessage;
  }

  public String getDetails() {
    return details;
  }

  public Instant getTimestamp() {
    return timePosted;
  }

  public String calculateTimeFromWhenPosted() {
    String messageFormatted = "";

    final Duration timeBetween = Duration.between(this.getTimestamp(), Instant.now());
    final long numDays = timeBetween.toDays();
    final long numHours = timeBetween.toHours();
    final long numMinutes = timeBetween.toMinutes();
    final long numSeconds = timeBetween.toSeconds();

    if (numDays > 0) messageFormatted = getFormatForTimeUnit(numDays, "day");
    else if (numHours > 0) messageFormatted = getFormatForTimeUnit(numHours, "hour");
    else if (numMinutes > 0) messageFormatted = getFormatForTimeUnit(numMinutes, "minute");
    else messageFormatted = getFormatForTimeUnit(numSeconds, "second");

    return String.format("(%s)", messageFormatted);
  }

  private String getFormatForTimeUnit(final long timeUnit, final String timeUnitSimpleName) {
    return String.format("%s %s ago", timeUnit, appendIfPlural(timeUnit, timeUnitSimpleName));
  }

  public String appendIfPlural(final long timeUnit, final String timeUnitSimpleName) {
    if (timeUnit < 0 || timeUnit == 1) return timeUnitSimpleName;
    else return timeUnitSimpleName + "s";
  }

  public String formatMessage() {
    return this.getDetails() + " " + calculateTimeFromWhenPosted();
  }

  public String formatMessageWithNameOfUser() {
    return this.getUserWhoPostedMessage().getName() + " - " + this.formatMessage();
  }

  public static final Comparator<Message> SORT_BY_TIME_STAMP =
      Comparator.comparing(Message::getTimestamp);
}

File: starter/java/src/main/kotlin/kata/Message.kt
package kata

import lombok.Data
import lombok.Getter
import java.time.Duration
import java.time.Instant

data class Message(
	private val userWhoPostedMessage: User,
	val details: String,
) {

	val timePosted: Instant = Instant.now()

	fun calculateTimeFromWhenPosted(): String {
		val timeBetween = Duration.between(this.timePosted, Instant.now())
		val numDays = timeBetween.toDays()
		val numHours = timeBetween.toHours()
		val numMinutes = timeBetween.toMinutes()
		val numSeconds = timeBetween.toSeconds()

		val messageFormatted = if (numDays > 0) getFormatForTimeUnit(numDays, "day")
		else if (numHours > 0) getFormatForTimeUnit(numHours, "hour")
		else if (numMinutes > 0) getFormatForTimeUnit(numMinutes, "minute")
		else getFormatForTimeUnit(numSeconds, "second")

		return "(${messageFormatted})"
	}

	private fun getFormatForTimeUnit(timeUnit: Long, timeUnitSimpleName: String) =
		"$timeUnit ${this.appendIfPlural(timeUnit, timeUnitSimpleName)} ago"

	fun appendIfPlural(timeUnit: Long, timeUnitSimpleName: String): String {
		return if (timeUnit < 0L || timeUnit == 1L) timeUnitSimpleName
		else "${timeUnitSimpleName}s"
	}

	fun formatMessage() = "$details ${calculateTimeFromWhenPosted()}"

	fun formatMessageWithNameOfUser() = "${this.userWhoPostedMessage.name} - ${this.formatMessage()}"

//	public fun <Message> Comparator test =

	class SortByTimeStamp : Comparator<Message> {
		override fun compare(o1: Message?, o2: Message?): Int {
			if (o1 == null || o2 == null) return 0
			return o1.timePosted.compareTo(o2.timePosted)
		}
	}
}
